composer/datasets/streaming/dataset.py

```diff
@@ -229,99 +230,191 @@ def _insert_shard_samples(self, shard: int, part_min_id: int, part_max_id: int)
                     np.random.shuffle(todo_ids)
             else:
                 if self._download_status == _DownloadStatus.IN_PROGRESS:
-                    self._downloaded_ids.reverse()
                     self._downloaded_ids.extend(new_ids)
-                    self._downloaded_ids.reverse()
                 for todo_ids in self._epoch_to_todo_ids.values():
                     todo_ids.reverse()
                     todo_ids.extend(new_ids)
                     todo_ids.reverse()
 
-    def download(self) -> None:
-        """Download and assimilate missing shards."""
+            # Note that we have loaded the shards.
+            for shard in shards:
+                self._has_shard[shard] = True
+
+    def _load_any_local_shards(self) -> List[int]:
```

Comment:
There's a few things going on in this function, which makes it harder to name and follow:

1. It creates the lock and initializes the download status
2. It determines what shards are already downloaded and what shards need to be downloaded
3. It loads already downloaded shards into the internal state
4. It shuffles the list of missing shards

Could this function be broken up -- perhaps into four private methods? I'm thinking that part 3 could re-use the same logic as when the download helper finishes a download